stdlib/camlinternalLazy.ml

```diff
@@ -59,83 +54,73 @@ external reset_to_lazy : Obj.t -> unit = "caml_lazy_reset_to_lazy" [@@noalloc]
+  (* In the common case, expect the tag to be [lazy_tag], but may be other tags
+     due to concurrent forcing of lazy values. *)
   match update_to_forcing (Obj.repr blk) with
   | 0 when only_val -> do_force_val_block blk
   | 0 -> do_force_block blk
-  | 1 -> raise Undefined
   | _ ->
-      (* We raise [RacyLazy] for all the other cases.
-       *
-       * If the return value was 2, we should raise RacyLazy. We had observed
-       * earlier in [force_gen] that the [blk] was not [Obj.forward_tag] and
-       * not short-circuited. But now we observe that the value has either been
-       * forwarded (return value 3) or forwarded and short-circuited (4). Both
-       * of these indicate a race. *)
-      raise RacyLazy
+      (* If the return value is 1, then the tag is [forcing_tag]. We are either
+         recursively forcing on the same domain, or concurrently forcing from
+         multiple domains. Raise [Undefined].
+
+         If the return value is 2, then the tag is [forward_tag]. In
+         [force_gen], we checked whether the tag was [forward_tag]. It has
+         changed since the last read. Hence, there is another domain which is
+         concurrently forcing this lazy. Raise [Undefined].
+
+         If the return value is 3, then the tag is not [forcing_tag] or
+         [forward_tag], which we checked in [force_gen]. The tag is not
+         [lazy_tag] or otherwise [update_to_forcing] would have succeeded and
+         returned [0]. The tag has changed from the expected [lazy_tag] due to
+         another domain concurrently, and the GC short circuiting the result.
+         Raise [Undefined].
+         *)
+      raise Undefined
 
 (* used in the %lazy_force primitive *)
 let force_lazy_block blk = force_gen_lazy_block ~only_val:false blk
 
-let try_force_gen_lazy_block ~only_val (blk : 'arg lazy_t) =
-  match update_to_forcing (Obj.repr blk) with
-  | 0 when only_val -> Some (do_force_val_block blk)
-  | 0 -> Some (do_force_block blk)
-  | 1 -> raise Undefined
-  | _ ->
-      (* Other cases indicate racy access to lazy for the same reasons outlined
-       * in [force_gen_lazy_block]. *)
-      None
-
 (* [force_gen ~only_val:false] is not used, since [Lazy.force] is
    declared as a primitive whose code inlines the tag tests of its
    argument, except when afl instrumentation is turned on. *)
-
 let force_gen ~only_val (lzv : 'arg lazy_t) =
   let x = Obj.repr lzv in
-  (* XXX No safe points start *)
   let t = Obj.tag x in
-  if t = Obj.forward_tag then (Obj.obj (Obj.field x 0) : 'arg)
-  (* XXX No safe points end *)
-  else if t <> Obj.lazy_tag && t <> Obj.forcing_tag then (Obj.obj x : 'arg)
+  (* START no safe points. If a GC occurs here, then the object [x] may be
+     short-circuited, and getting the first field of [x] would get us the wrong
+     value. Luckily, the compiler does not insert GC safe points at this place,
+     so it is ok. *)
+  if t = Obj.forward_tag then
+    (Obj.obj (Obj.field x 0) : 'arg)
```

Comment:
I am trying to convince myself that this is correct according to OCaml's memory model, but I do not see what ensures that the field is seen as set despite the tag being seen as a forward_tag. Don't you need a load/load barrier before reading the field, or reading the tag with an acquire load, or something stronger? (To my knowledge, OCaml's memory model does not have relaxed atomics yet, so furthermore only a plain atomic read of the tag makes sense here yet.) What makes this correct (and the similar code inlined by lambda)?

Reply:
My understanding is that the memory model gives a message-passing semantics to non-atomic writes+reads that are synchronized through an atomic at another location.

In the operational semantics of the memory-model paper: if I write to an atomic location, the location's frontier is updated with my frontier, including non-atomic writes that I did before (in program order). Then when the other domain reads from this atomic location, its frontier is updated by my frontier, so it also sees my non-atomic writes to other locations.

Reply:
Sure, but for this you need an atomic read as I wrote

Reply:
We should be using C11 memory model to be reasoning about this as the tag update `update_to_forward` and `Obj.tag` are both external calls implemented in C. `update_to_forward` uses `atomic_compare_and_exchange_strong`, but `Obj.tag`, which calls `caml_obj_tag` uses non-atomic reads. My suggestion is to replace the `Tag_val` in `obj.c:obj_tag` with an acquire load. This should be sufficient to ensure that the message-passing semantics works between the tag update and the field write. @stedolan?

Reply:
Updated the tag read to be load acquire. See https://github.com/ocaml-multicore/ocaml-multicore/pull/760/commits/57137194f2be0caec3697a5980d0851763649d34#diff-5dfaf7e6dc66b87cf561b38c634d235d2e5d747c2a9e27df18f290bc223533cfR43-R44

Reply:
I'm a bit nervous about the performance implications of making `Obj.tag` an atomic read all over the place, "just" for the need of lazy values -- I hope it is uncommon to concurrently change a value tag, and it's not clear that we should pay the price for all tag lookups. Wouldn't it be more prudent to expose an `obj_tag_acquire` variant (or `obj_tag_atomic`) and use it?

Reply:
Note: `Obj.tag` is already known as a not-so-efficient way to read tags, but currently there is no faster/simpler primitive (so it remains used in all tag-manipulating OCaml code). And whether or not `Obj.tag` is slow, I think that it may be beneficial to force ourselves to express the intent that the block we are accessing may be concurrently mutated.

Reply:
Added `caml_obj_tag_atomic` for this. I've unconditionally used this in lazy implementation, as I'd rather err on the side of safety than performance for lazy for MVP.

Reply:
Don't forget that there is similar code in `lambda/matching.ml` which will need an acquire load as well.

Alternatively, you only need a load/load barrier (e.g. `dmb ishld` on aarch64) before reading the field, which removes the synchronisation cost for short-circuited blocks (though it is likely very small).

Eventually it would be nice to have acquire/release atomics in the OCaml memory model, or barriers to reason about non-atomic locations like here, because it looks complicated to reason about mixing two memory models (and especially that this does not introduce any catch-fire semantics).

Note that since the header is used as an atomic location by the runtime, the tag must always be read from C with at least a `relaxed` atomic read (or else it can catch fire).

Reply:
To clarify, my comment about mixing memory models concerns the inlined forcing code at `matching.ml`.

Acquire/release fences are the most natural in my alternative solution at https://gitlab.com/gadmm/stdlib-experiment/-/blob/master/other/sync_lazy.ml (which also avoids mutating the tag concurrently); and an OCaml version of a release fence might only need a `dmb st` fence.

Reply:
I've already put in the inlined fixes for `lambda/matching.ml`. See https://github.com/ocaml-multicore/ocaml-multicore/pull/760/files#diff-059d6544bdc63e5f278818ea9f50bef7aa3f855c4d5ce6fd726abee18b187f55R1860-R1861.

Reply:
I'm not comfortable using raw fences in the OCaml. Our PLDI'18 paper does not include them, and we'd have to extend the work to talk about the guarantees, and in particular, preserving LDRF and their interaction with the rest of the memory model. I wouldn't think of adding fences lightly. 

I would like to aim for a working implementation in this PR, even if it is not the most efficient, and save the optimisations for after multicore MVP is merged with OCaml.

Reply:
To be more precise, `inline_lazy_force_switch` in `matching.ml` does not use `prim_obj_tag` so I assume that it makes an unsynchronized read. This is the one used to compile native code (for efficiency). Possible solutions:
1. replace it with a call to `inline_lazy_force_cond` (remove optimisation)
2. add an argument to Lswitch to read-acquire the tag (heavy solution with a single use-case)
3. add a load-load fence before reading the field (only minimal working solution of a fence for internal use needed in a first time, more carefully design can come later if programmers need it)
4. use my variant above that stores the forward bit in the thunk field rather than with the tag, and atomically read this field.

If you do not want a fence, then 1. and 4. are the simplest I see (and we could do 1. first and experiment with 4. later).

Reply:
> Added caml_obj_tag_atomic for this.

Apologies for yet another dissenting opinion, but I don't like the idea of having multiple versions of C functions that differ only in the memory model used.

First, it's a software engineering nightmare: code duplication + tiny changes...  We already have enough smelly C code to maintain, thank you very much for not adding more.

Second, the performance degradation that @gasche is worried about has not been demonstrated.  It will be null on x86 (all loads have acquire behaviors) and (in my guess) negligible on other platforms, since `caml_obj_tag` is quite expensive already.

Third, I think the C primitives should be made to fit the Multicore OCaml memory model, and not the other way around.  Meaning that if a C11 nonatomic load in a C primitive is a source of concern, it makes perfect sense to make more memory coherent for all uses.  And perhaps other primitives from Obj should be scrutinized from this angle too.

Reply:
Here is a proposal for `inline_lazy_force_switch`:

```diff
diff --git a/lambda/matching.ml b/lambda/matching.ml
index 985f86e25..d996489d2 100644
--- a/lambda/matching.ml
+++ b/lambda/matching.ml
@@ -1943,12 +1943,12 @@ let inline_lazy_force_switch arg loc =
         ( Lprim (Pisint, [ varg ], loc),
           varg,
           Lswitch
-            ( varg,
-              { sw_numconsts = 0;
-                sw_consts = [];
-                sw_numblocks = 256;
+            ( Lprim (Pccall prim_obj_tag, [ varg ], loc),
+              { sw_numblocks = 0;
+                sw_blocks = [];
+                sw_numconsts = 256;
                 (* PR#6033 - tag ranges from 0 to 255 *)
-                sw_blocks =
+                sw_consts =
                   [ (Obj.forward_tag, Lprim (Pfield(0, Pointer, Mutable),
                                              [ varg ], loc));
```

Unfortunately, the compiler generates not-so-good code in this case (it handles poorly a sparse switch with only integer constants). I have some matching-specific ideas to improve it, but it should be discussed with matching experts and is somewhat irrelevant to the current PR discussion. I believe that this version is better than reverting to `inline_lazy_force_cond`, so it would be a good first step for that function.

Reply:
@xavierleroy I'm nervous because `Obj.tag`-like functionality is used for pattern-matching, so actually quite heavily used in OCaml programs. Something smells bad about the idea that the dirty tag tricks we use for lazy thunks spreads overhead into all OCaml programs, even those that never use Lazy. (Maybe the cost is neglectible but, if we are going to pay it, we should measure it!)

More importantly, we should think hard about whether we can actually safely mutate tags in this way for lazy thunks, and which parts of the runtime are affected. Adding a barrier to `Obj.tag` is not going to make the problem go away. Having two different primitives, and forcing us to think about which one to use where, is a way to record our decisions/understanding about this problem. Maybe in the end we would find out that we only use the atomic version, and we can simplify again, but in the meantime we should have a clear way to tell which parts of the codebase have to account for concurrent tag modifications, and which do not.

Reply:
> I'm nervous because Obj.tag-like functionality is used for pattern-matching

Can you be more precise?  Does the pattern-matching compiler generate calls to `caml_obj_tag` or similar C primitives?  The "get tag" operations produced to implement Lswitch, etc, are under the control of the OCaml compiler, so they obey the Multicore OCaml memory model.

Reply:
I reviewed the use of `Obj.tag` in the compiler codebase; @xavierleroy is correct (of course!) that there is mostly nothing performance-sensitive there, and that adding an atomic load to all `Obj.tag` users would be fine. The use-cases of `caml_obj_tag`  in `lambda/matching.ml` come from the compilation of `lazy` patterns, and I guess it's okay if those suffer from overhead. The compilation of non-lazy patterns uses a different `get_tag` operation.

> The "get tag" operations produced to implement Lswitch, etc, are under the control of the OCaml compiler, so they obey the Multicore OCaml memory model.

Right but, if I understand correctly, as @gadmm pointed out the Multicore OCaml memory model says little about what you get if you non-atomically read a field (as `get_tag` does) that was atomically-written to by a Lazy mutation (as `Lazy.force` does). My understanding is that calling `get_tag` to learn that a block has one of the lazy tags (Lazy, Forcing, Forward) *would* be racy and not guarantee anything about the block arguments. We are safe because we never do this -- and my unboxed-constructors work is also safe, as `Lazy.t` is known to have any shape due to Forward shortcutting.

Reply:
It is true that the memory model does not say what happens when treating an atomic location as non-atomic (maybe the model can trivially be extended to say what happens in this case), but I was more pointing out that would be racy anyway (and the alternative of a strong atomic load is not nice).